feat(web): Show lens model in the asset viewer detail panel

bxtdvd opened the pull request:
This adds the lens model information to the EXIF area of the asset view info detail panel.  This is useful because the display of the focal length alone doesn't necessarily tell you what lens you used for a particular shot.

Additionally, I've made both the camera make/model and the lens model clickable links. These go to the search page with a query of either that camera make + camera model (if you click on the camera make/model), or the lens model (if you click on the lens model).

This is based on and inspired by the link part of https://github.com/immich-app/immich/pull/15049, which has been super useful to me.

This is my first PR and I'm a Svelte newbie, so please let me know if there's anything I did wrong. I ran the code through all the npm checks.

![Screenshot 2025-01-19 at 8 59 41 PM](https://github.com/user-attachments/assets/783ab479-08ae-4525-abf5-23cf27c312cc)

bxtdvd committed:
Adds lens details to the asset viewer

alextran1502 commented:
Thank you for the PR. I have two concerns about the changes that I don't know how to address.

1. There is no UI indicator that these two items are clickable.
2. the lens info will be long for many iPhone users. It will be displayed as two lines of text. It looks a bit cluster

I came up with an attempt for solutions, but I am still not very happy with it wrt the density of information.

<img width="381" alt="image" src="https://github.com/user-attachments/assets/95bec96b-16fa-4835-831c-2a1689a410b8" />

<img width="385" alt="image" src="https://github.com/user-attachments/assets/90f190c2-1828-4a18-9672-f7ab00d2f855" />

chrneu commented:
Would it be possible to crop the second line out and display the full length as ALT attribute? In the end the content of the second line from the iphone is duplicated information anyway.

bxtdvd commented:
I like the idea of clamping the line and providing the remaining text in the hover. The most important info does usually come first.
<img width="545" alt="Screenshot 2025-01-20 at 8 21 05 AM" src="https://github.com/user-attachments/assets/0b8569a2-f8a7-433b-bcf9-773ea071aff2" />

I did initially think about doing this with a search icon, but was concerned that might be too visually noisy for a somewhat niche feature. I'm not a UI/UX designer, so I'm certainly open to any other ideas here, too.

Showing the lens model is the most important part of this feature. If linking to the search makes things too clunky for the UI, then I can remove that.

Snowknight26 commented on web/src/lib/components/asset-viewer/detail-panel.svelte:
```diff
@@ -410,7 +425,31 @@
+                  title={$t('search_for_lens')}
+                  class="hover:dark:text-immich-dark-primary hover:text-immich-primary"
+                >
+                  {asset.exifInfo.lensModel || ''}
```
Don't think this logical OR is needed because of the if on 439.

Snowknight26 commented on web/src/lib/components/asset-viewer/detail-panel.svelte:
```diff
@@ -410,7 +425,31 @@
         <div>
-          <p>{asset.exifInfo.make || ''} {asset.exifInfo.model || ''}</p>
+          <p>
+            <a
+              href={getCameraSearchHref(asset)}
+              title={$t('search_for_camera')}
+              class="hover:dark:text-immich-dark-primary hover:text-immich-primary"
+            >
+              {asset.exifInfo.make || ''}
+              {asset.exifInfo.model || ''}
+            </a>
```
If neither the make nor model are set, will this still render the `a`? Seems like you'd want to do the check higher up, similar to line 439.

Snowknight26 commented on web/src/lib/components/asset-viewer/detail-panel.svelte:
```diff
@@ -142,6 +142,21 @@
+    const cameraSearchUrl = new URL(AppRoute.SEARCH, globalThis.location.href);
+    cameraSearchUrl.searchParams.set(
+      QueryParameter.QUERY,
+      `{"make":"${asset.exifInfo?.make}","model":"${asset.exifInfo?.model}"}`,
```
This'll not work if the make or model have double quotes in them. It's probably best to use `JSON.stringify` for the entire value. Same concerns with `getLensModelSearchHref`.

I also think this'll give you unexpected results if the make is unset or if the model is unset, since converting `undefined` to a string gives you the string literal `"undefined"`.

Additionally, you could probably deduplicate the two get*Href functions further since everything but the `value` parameter of the `set` call is the same.

Does a helper function already exist to be able to specify the search URL? Quick digging reveals `getMetadataSearchQuery`, so I imagine you'd want to use that and/or whatever code currently calls this function.

dclobato commented:
> This adds the lens model information to the EXIF area of the asset view info detail panel

Nice!  Looking forward to have this PR merged!  🤞

alextran1502 commented:
> I like the idea of clamping the line and providing the remaining text in the hover. The most important info does usually come first. <img alt="Screenshot 2025-01-20 at 8 21 05 AM" width="545" src="https://private-user-images.githubusercontent.com/30203091/404946341-0b8569a2-f8a7-433b-bcf9-773ea071aff2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0OTc4MzcsIm5iZiI6MTczNzQ5NzUzNywicGF0aCI6Ii8zMDIwMzA5MS80MDQ5NDYzNDEtMGI4NTY5YTItZjhhNy00MzNiLWJjZjktNzczZWEwNzFhZmYyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIxVDIyMTIxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkNzM0Mjk2M2RiZDdiNjVkM2JlNDUwMTg2NTY2MDNhYjJlNDc2ZDJjZDM1MzE4Y2U2ODQ3YjlmYmNmNWE0NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BWbkrPjTfMnvNKBLNHjtdVF1bFqgHqPtc5npSsI31Sg">
> 
> I did initially think about doing this with a search icon, but was concerned that might be too visually noisy for a somewhat niche feature. I'm not a UI/UX designer, so I'm certainly open to any other ideas here, too.
> 
> Showing the lens model is the most important part of this feature. If linking to the search makes things too clunky for the UI, then I can remove that.

Let's implement the elapsed text here and not add the search icon but keep the search feature. We can find a way to make an indicator nicer later

bxtdvd commented:
Thanks Alex. I was away on travel yesterday, but I'll update the code with this — as well as with the changes that Snowknight26 suggested — this evening.

bxtdvd committed:
Update lens detail search links

bxtdvd commented:
I've updated the code with the lens model clamped to 1 line. 

I've also simplified the implementation a bit by using the existing getMetadataSearchQuery helper function. I also added additional error checking for situations where either camera make or model is blank. I tested this with some photos where I removed make or model with exiftool.

Thanks @Snowknight26 for the review and your suggestions on how I could improve the code!

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into lens-details

alextran1502 reviewed (approved):
Thanks much! looks very good
